stdlib/public/Concurrency/TaskGroup.cpp

```diff
@@ -1739,21 +1739,52 @@ PollResult TaskGroupBase::tryEnqueueWaitingTask(AsyncTask *waitingTask) {
   bool haveRunOneChildTaskInline = false;
 
   reevaluate_if_TaskGroup_has_results:;
-  auto assumed = statusMarkWaitingAssumeAcquire();
```

Comment:
The change here is that we don't mark and undo, but only mark when we indeed decide that we'll be waiting.

This is all still done with a lock, but the dance is more correct anyway.